### DISCUSSION TITLE
Consistent behavior accross DB Engines

### DISCUSSION THREAD

**Discussion by stackcoder** [firefly-iii/firefly-iii#9324] [Firefly III]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [x] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/explanation/more-information/what-its-not/)
- [x] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

Sorry for re-opening this controversial topic. I'm pretty sure everyone agrees on this:

A Search that requires matching the exact letter case is unexpected and not the biggest help.

Of course, I'm aware of those many closed and locked issues #2552, #8854, #8187, #7769, #4979, #4961, and the FAQ. Personally, I'm an everyday user of this application, and just migrated to PostgreSQL, but this little detail doesn't satisfy me. Using the same DB engine for all other applications saves me time and valuable resources. All things on my computer are open source, this is definitely something we can fix! And I want to contribute constructively.

### Solution

I've looked into some threads, first I tried to "configure it away" using the Postgres V12 `collation` feature as pointed out by @chevdor in #4961.

```sql
CREATE COLLATION english_ci (
  PROVIDER = 'icu',
  LOCALE = 'en-US@colStrength=secondary',
  DETERMINISTIC = TRUE
);

ALTER TABLE transaction_journals ALTER description TYPE character varying(1024) COLLATE english_ci;
```

Unfortunately, nondeterministic collations are [not supported](https://www.cybertec-postgresql.com/en/case-insensitive-pattern-matching-in-postgresql/) for LIKE and will result in error.

So we have various options still to go:
1. [Patch](https://www.postgresql.org/message-id/700d2e86-bf75-4607-9cf2-f5b7802f6e88%40eisentraut.org)ing PostgreSQL
Maintenance is not worth it. Plus, any user-made change to the DB is not covered by automated testing and can cause the next update to fail.

2. Make Doctrine ORM render `LIKE`as `ILIKE` on PostgreSQL
This way PostgreSQL would behave more like MariaDB. 

3. Extend Doctrine ORM with a custom `ILIKE FunctionNode` and explicitly define which comparison should be case insensitive.
Might be technologically the most correct way to implement, without performance impacts, but requires some effort.

4. Use `lower()` inside the query
The most pragmatic approach was already suggested by @arnvald. My candidate for hot fixing my system without fear of feature incompatibility.

Can we agree, that the application is the only thing that knows if a comparison should be case-sensitive or not? Neither the ORM nor the DB Engine can understand what the user expects. In my experience, most applications go with Option 3 or 4. I also had a [very brief look](https://github.com/search?q=repo%3Anextcloud%2Fserver%20ilike&type=code) in NextCloud one of the most popular PHP applications out there.

### What are alternatives?

The worst option ever would be entirely dropping PostgreSQL support and migrating back to MariaDB. Integrating something like ElasticSearch is more an overkill than a simple solution and should be optional. I would be happy to have case-insensitive string contains again.

**Comment by JC5** [comment]
I for one am very interested in a solution that actually works across all engines.

**Comment by stackcoder** [comment]
### Good News: The perfect solution is already in  Laravel

Since the shipped Laravel version already contains the [new whereLike clause](https://github.com/laravel/framework/pull/52147) we just need to replace `where` with `whereLike`.

I can confirm applying following patch to my environment improves the search drastically:

```diff
--- a/app/Helpers/Collector/GroupCollector.php
+++ b/app/Helpers/Collector/GroupCollector.php
@@ -944,7 +944,7 @@
                     static function (EloquentBuilder $q1) use ($array): void {
                         foreach ($array as $word) {
                             $keyword = sprintf('%%%s%%', $word);
-                            $q1->where('transaction_journals.description', 'LIKE', $keyword);
+                            $q1->whereLike('transaction_journals.description', $keyword);
                         }
                     }
                 );
```

Sure, there are some more occurrences which should be replaced. For now my quick code search shows 63 results in 23 files. 

*Sorry for mention Doctrine above - Doctrine is not (anymore) referenced by Composor.*

**Reply by JC5** [reply]
Nice, finally, that should be a quick fix. Thanks for sharing this!

**Reply by JC5** [reply]
All instances have been updated in the develop branch. A new development release will go out next Monday. Thanks again!

**Reply by stackcoder** [reply]
That's impressively fast, thank you!

**Reply by JC5** [reply]
Hah, thanks. But equally thank _you_, I'm very happy you found this solution.

**Reply by matejdro** [reply]
Thank you a lot! At last this annoying issue is solved!
